src/main/java/org/candlepin/subscriptions/tally/roller/DailySnapshotRoller.java

```diff
@@ -0,0 +1,136 @@
+    }
+
+    @Transactional(value = "inventoryTransactionManager", readOnly = true)
+    public List<ProductUsageCalculation> calculateMaxValuesFromInventory(List<String> accounts) {
+        Map<String, List<InventoryHost>> hostsByAccount =
+            inventoryRepository.findByAccountIn(accounts)
+            .collect(Collectors.groupingBy(InventoryHost::getAccount));
+
+        List<ProductUsageCalculation> calculations = new LinkedList<>();
+        for (Entry<String, List<InventoryHost>> hostEntry : hostsByAccount.entrySet()) {
+            // NOTE: When we start counting for other products, consider adding additional
+            //       calculators to a list and applying the host to each.
+            String accountNumber = hostEntry.getKey();
+            ProductUsageCalculation rhelUsageCalculation = new ProductUsageCalculation(accountNumber,
+                this.product);
+            hostEntry.getValue().forEach(host -> {
+                NormalizedFacts facts = factNormalizer.normalize(host);
+                if (facts.getProducts().contains(rhelUsageCalculation.getProductId())) {
+                    rhelUsageCalculation.addCores(facts.getCores() != null ? facts.getCores() : 0);
+                    rhelUsageCalculation.addInstance();
+                }
+
+                // Validate and set the owner.
+                String owner = facts.getOwner();
+                if (owner == null) {
+                    // Don't set null owner as it may overwrite an existing value.
+                    // Likely won't happen, but there could be stale data in inventory
+                    // with no owner set.
+                    return;
+                }
+
+                String currentOwner = rhelUsageCalculation.getOwner();
+                if (currentOwner != null && !currentOwner.equalsIgnoreCase(owner)) {
+                    throw new IllegalStateException(
+                        String.format("Attempt to set a different owner for an account: %s:%s",
+                            currentOwner, owner));
+                }
+                rhelUsageCalculation.setOwner(owner);
+            });
+            log.debug("Found {} cores of RHEL for account: {}", rhelUsageCalculation.getTotalCores(),
+                rhelUsageCalculation.getAccount());
+            calculations.add(rhelUsageCalculation);
+        }
+        return calculations;
+    }
+
+    private void produceDailySnapshots(List<String> accounts, List<ProductUsageCalculation> calculations) {
+
+        Map<String, AccountMaxValues> maxValues = calculations.stream()
+            .map(this::createMaxValuesFromCalculation)
+            .collect(Collectors.toMap(AccountMaxValues::getAccountNumber, Function.identity()));
+
+        log.debug("Fetching existing daily snapshots for {} accounts.", accounts.size());
+
+        Map<String, TallySnapshot> existingRhelSnapsForToday =
+            tallyRepo.findByAccountNumberInAndProductIdAndGranularityAndSnapshotDateBetween(
```

Comment:
[From the Spring Data docs](https://docs.spring.io/spring-data/jpa/docs/current/reference/html/#repositories.query-streaming), "a Stream potentially wraps underlying data store-specific  resources and must, therefore, be closed after usage. You can either  manually close the Stream by using the close() method or by using a Java 7 try-with-resources block."

So, I think something like

```java
try (Stream<TallySnapshot> s = tallyRepo.findByAccountNumberInAndProductIdAndGranularityAndSnapshotDateBetween(accounts, this.product, TallyGranularity.DAILY, clock.startOfToday(), clock.endOfToday())) {
    s.collect(Collectors.toMap(TallySnapshot::getAccountNumber, Function.identity())
}

```

Reply:
How's something like this?

```
    private void produceDailySnapshots(List<String> accounts, List<ProductUsageCalculation> calculations) {

        Map<String, AccountMaxValues> maxValues = calculations.stream()
            .map(this::createMaxValuesFromCalculation)
            .collect(Collectors.toMap(AccountMaxValues::getAccountNumber, Function.identity()));

        log.debug("Fetching existing daily snapshots for {} accounts.", accounts.size());

        try(Stream<TallySnapshot> existingSnapStream =
            tallyRepo.findByAccountNumberInAndProductIdAndGranularityAndSnapshotDateBetween(
            accounts, this.product, TallyGranularity.DAILY, clock.startOfToday(), clock.endOfToday())) {
            Map<String, TallySnapshot> existingRhelSnapsForToday = existingSnapStream.collect(
                Collectors.toMap(TallySnapshot::getAccountNumber, Function.identity()));

            log.debug("Found {} existing snapshots for today.", existingRhelSnapsForToday.size());

            updateSnapshots(existingRhelSnapsForToday, maxValues, TallyGranularity.DAILY);
        }

    }
```

Reply:
or should it be closed right after collect?

Reply:
Looking at it, it should probably get closed after the collect.

Reply:
@awood something like this.
```
    private void produceDailySnapshots(List<String> accounts, List<ProductUsageCalculation> calculations) {

        Map<String, AccountMaxValues> maxValues = calculations.stream()
            .map(this::createMaxValuesFromCalculation)
            .collect(Collectors.toMap(AccountMaxValues::getAccountNumber, Function.identity()));

        log.debug("Fetching existing daily snapshots for {} accounts.", accounts.size());

        Map<String, TallySnapshot> existingSnapsForToday;
        try(Stream<TallySnapshot> existingSnapStream =
            tallyRepo.findByAccountNumberInAndProductIdAndGranularityAndSnapshotDateBetween(
            accounts, this.product, TallyGranularity.DAILY, clock.startOfToday(), clock.endOfToday())) {
            existingSnapsForToday = existingSnapStream.collect(
                Collectors.toMap(TallySnapshot::getAccountNumber, Function.identity()));
            log.debug("Found {} existing snapshots for today.", existingSnapsForToday.size());
        }

        updateSnapshots(existingSnapsForToday, maxValues, TallyGranularity.DAILY);
    }
```

Reply:
Yeah, but I think we should DRY that up some.

Can we have something like a `generateTallyMap` function in the base class that does most of this and just externalizes the granularity to another method `getRollerGranularity`?  The granularity looks like the only thing that varies across the implementations of the Rollers.

Reply:
DONE - I did it a little differently than you suggested due to some other constraints like dates. Let me know what you think.